README.md
# slim
slim login manager and locker

#GalliumOS
This is a fork of SLiM for GalliumOS


ChangeLog
1.3.7 - 2015.10.22
        * Changes for GalliumOS
	* Added XRandr support for resizing slimlock
        * Disabled tty locking by default
        * Run session stop command even if login command fails
        * Various fixes and changes

1.3.6 - 2013.10.01
	* Merge slimlock.
	* Add support ^H (like backspace).
	* And fix some bugs.

1.3.5 - 2012.12.31
	* Support UTF8 string.
	* Add systemd service.
	* And fix some bugs.

1.3.4 - 2012.06.26
	* Replaced stderr writes function.
	* Fix numlock control.
	* Fix build with GLUT.
	* Fix PAM authentication.

1.3.3 - 2012.02.22
	* Change build system to CMake.
	* Add support ConsoleKit.
	* Fix some bugs....

1.3.2 - 2010.07.08
    * Add support xauth secret.
    * Add xnest_debug mode.

1.3.1 - 2008.09.26
    * Added focus_password config option for focusing password
      automatically when default_user is enabled
    * Added auto_login option
    * Fixed uninitialized daemonmode, see
      http://www.freebsd.org/cgi/query-pr.cgi?pr=114366
    * Fixed maximum length for password
    * Introduced customization options for session text:
      font, colors, position, shadows.

1.3.0 - 2006.07.14
    * Added PAM support by Martin Parm
    * Fixed segfault on exit when testing themes. Thanks
      to Darren Salt & Mike Massonnet
    * Fixed vt argument detection, thanks to Henrik Brix Andersen
    * Corrected reference to input_color in the default theme
    * Fixed default shell setting
    * Fix segfault when calling XCloseDisplay(NULL); thanks Uli Schlachter

1.2.6 - 2006.09.15
    * Bug #008167: Update pid when in daemon mode
    * Fixed warnings when compiling with -Wall. Thanks to
      KIMURA Masaru
    * Fixed major memory leaks with repeated login (bug #007535)

1.2.5 - 2006.07.24
    * hiding of the cursor is now an option (disabled
      by default) since some WMs does not re-initialize
      the root window cursor.
    * The X server is restarted when the user logs out.
      This fixes potential security issues with user-launched
      apps staying attached to the root window after logout.
    * Bug #7432 : Added proper Xauth authentication: the X server
      is started with the -auth option and the user who logs
      in has his .Xauthority file initializated.

1.2.4 - 2006.01.18
    * Added commands for session start and stop
      (i.e. for session registering)
    * Added automatic numlock on/off option
    * Support for numpad Enter key
    * Restored support for daemon option in the config
      file.
    * Lock file now uses process id, no more false
      locking (thanks to Tobias Roth)

1.2.3 - 2005.09.11
    * Added FreeBSD, NetBSD, OpenBSD support
    * Replaced autotools with plain makefile(s)
    * Added 'suspend' command (untested, we don't use it)
    * Added support for %theme variable in login command
	
1.2.2 - 2005.05.21
    * fix panel drawing on screens <= 1024x768
    * Don't start X server unless valid theme found
    * revert to 'default' of invalid theme specified
    * try all themes from a set if one doesn't work
    
1.2.1 - 2005.05.17
    * draw input directly on panel

1.2.0 - 2005.05.16
    * added theme preview (slim -p /path/to/theme)
    * added JPEG support for panel image
    * added 'center' background type and 'background_color' option
    * added text shadow
    * added warning when execution of login command fails
    * Fix login failure when no shell specified in /etc/passwd
    * Print error when login command execution fails
    * add XNEST_DEBUG ifdef's to allow for easy debugging
    * Add support for Ctrl-u and Ctrl-w
    * Add 'vt07' to server arguments if not already specified
    * Removes daemon option from the config file. Use slim -d
    * Allow 'current_theme' to be a set of themes, choose randomly
    * Change default theme

1.1.0 - 2004.12.09
    * error messages for X11 apps are no longer redirected
      to the log file
    * fixed text position for default theme
    * added configurable shutdown and reboot messages
    * separated 'Enter username' and 'Enter password' messages
      position.
    * due to the previous two points, the theme format has 
      slightly changed
    
1.0.0 - 2004.12.07
    * First public SLiM release


cfg.h
/* SLiM - Simple Login Manager
   Copyright (C) 2004-06 Simone Rota <sip@varlock.com>
   Copyright (C) 2004-06 Johannes Winkelmann <jw@tks6.net>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
*/

#ifndef _CFG_H_
#define _CFG_H_

#include <string>
#include <map>
#include <vector>

#define INPUT_MAXLENGTH_NAME	30
#define INPUT_MAXLENGTH_PASSWD  50

#define CFGFILE SYSCONFDIR"/slim.conf"
#define THEMESDIR PKGDATADIR"/themes"
#define THEMESFILE "/slim.theme"

class Cfg {

public:
	Cfg();
	~Cfg();

	bool readConf(std::string configfile);
	std::string parseOption(std::string line, std::string option);
	const std::string& getError() const;
	std::string& getOption(std::string option);
	int getIntOption(std::string option);
	std::string getWelcomeMessage();

	static int absolutepos(const std::string &position, int max, int width);
	static int string2int(const char *string, bool *ok = 0);
	static void split(std::vector<std::string> &v, const std::string &str, 
					  char c, bool useEmpty=true);
	static std::string Trim(const std::string &s);

    std::pair<std::string,std::string> nextSession();

private:
	void fillSessionList();

private:
	std::map<std::string,std::string> options;
	std::vector<std::pair<std::string,std::string> > sessions;
	int currentSession;
	std::string error;
};

#endif /* _CFG_H_ */


panel.h
/* SLiM - Simple Login Manager
   Copyright (C) 1997, 1998 Per Liden
   Copyright (C) 2004-06 Simone Rota <sip@varlock.com>
   Copyright (C) 2004-06 Johannes Winkelmann <jw@tks6.net>
   Copyright (C) 2013 Nobuhiro Iwamatsu <iwamatsu@nigauri.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
*/

#ifndef _PANEL_H_
#define _PANEL_H_

#include <X11/Xlib.h>
#include <X11/keysym.h>
#include <X11/Xft/Xft.h>
#include <X11/cursorfont.h>
#include <X11/Xmu/WinUtil.h>
#include <sys/wait.h>
#include <stdlib.h>
#include <signal.h>
#include <iostream>
#include <string>

#ifdef NEEDS_BASENAME
#include <libgen.h>
#endif

#include "switchuser.h"
#include "log.h"
#include "image.h"

struct Rectangle {
	int x;
	int y;
	unsigned int width;
	unsigned int height;

	Rectangle() : x(0), y(0), width(0), height(0) {};
	Rectangle(int x, int y, unsigned int width,
		  unsigned int height) :
		x(x), y(y), width(width), height(height) {};
	bool is_empty() const {
		return width == 0 || height == 0;
	}
};

class Panel {
public:
	enum ActionType {
		Login,
		Lock,
		Console,
		Reboot,
		Halt,
		Exit,
		Suspend
	};

	enum FieldType {
		Get_Name,
		Get_Passwd
	};

	enum PanelType {
		Mode_DM,
		Mode_Lock
	};

	Panel(Display *dpy, int scr, Window root, Cfg *config,
	      const std::string& themed, PanelType panel_mode);
	~Panel();
	void OpenPanel();
	void ClosePanel();
	void ClearPanel();
	void WrongPassword(int timeout);
	void Message(const std::string &text);
	void Error(const std::string &text);
	void EventHandler(const FieldType &curfield);
	std::string getSession();
	ActionType getAction(void) const;

	void Reset(void);
	void ResetName(void);
	void ResetPasswd(void);
	void SetName(const std::string &name);
	const std::string& GetName(void) const;
	const std::string& GetPasswd(void) const;
	void SwitchSession();
	int MonitorsChanged();

private:
	Panel();
	void Cursor(int visible);
	unsigned long GetColor(const char *colorname);
	void OnExpose(void);
	void EraseLastChar(string &formerString);
	bool OnKeyPress(XEvent& event);
	void ShowText();
	void ShowSession();
	bool CheckXRandr();
	void SlimDrawString8(XftDraw *d, XftColor *color, XftFont *font,
			     int x, int y, const std::string &str,
			     XftColor *shadowColor,
			     int xOffset, int yOffset);
	void UpdateRootWindowSize();
	Rectangle GetPrimaryViewport();
	void ApplyBackground(Rectangle = Rectangle());

	/* Private data */
	PanelType mode; /* work mode */
	Cfg *cfg;
	Window Win;
	Window Root;
	Display *Dpy;
	int Scr;
	int X, Y;
	GC TextGC;
	GC WinGC;
	XftFont *font;
	XftColor inputshadowcolor;
	XftColor inputcolor;
	XftColor msgcolor;
	XftColor msgshadowcolor;
	XftFont *msgfont;
	XftColor introcolor;
	XftFont *introfont;
	XftFont *welcomefont;
	XftColor welcomecolor;
	XftFont *sessionfont;
	XftColor sessioncolor;
	XftColor sessionshadowcolor;
	XftColor welcomeshadowcolor;
	XftFont *enterfont;
	XftColor entercolor;
	XftColor entershadowcolor;
	ActionType action;
	FieldType field;
	//Pixmap   background;
	
	/* Username/Password */
	std::string NameBuffer;
	std::string PasswdBuffer;
	std::string HiddenPasswdBuffer;

	/*Xrandr stuff */
	int randr_event;
	int monitors_changed;

	/* screen stuff */
	Rectangle viewport;
	
	/* Configuration */
	int input_name_x;
	int input_name_y;
	int input_pass_x;
	int input_pass_y;
	int inputShadowXOffset;
	int inputShadowYOffset;
	int input_cursor_height;
	int welcome_x;
	int welcome_y;
	int welcome_shadow_xoffset;
	int welcome_shadow_yoffset;
	int session_shadow_xoffset;
	int session_shadow_yoffset;
	int intro_x;
	int intro_y;
	int username_x;
	int username_y;
	int username_shadow_xoffset;
	int username_shadow_yoffset;
	int password_x;
	int password_y;
	std::string welcome_message;
	std::string intro_message;

	/* Pixmap data */
	Pixmap PanelPixmap;

	Image *image;

	/* For thesting themes */
	bool testing;
	std::string themedir;

	/* Session handling */
	std::string session_name;
	std::string session_exec;
};

#endif /* _PANEL_H_ */


slimlock.cpp
/* slimlock
 * Copyright (c) 2010-2012 Joel Burget <joelburget@gmail.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 */

#include <cstdio>
#include <cstring>
#include <algorithm>
#include <sys/types.h>
#include <sys/ioctl.h>
#include <linux/vt.h>
#include <X11/keysym.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/extensions/dpms.h>
#include <security/pam_appl.h>
#include <err.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <errno.h>
#include <sys/file.h>
#include <fcntl.h>

#include "cfg.h"
#include "util.h"
#include "panel.h"

#undef APPNAME
#define APPNAME "slimlock"
#define SLIMLOCKCFG SYSCONFDIR"/slimlock.conf"

using namespace std;

void setBackground(const string& themedir);
void InitWindow();
void CloseWindow();
void HideCursor();
bool AuthenticateUser();
static int ConvCallback(int num_msgs, const struct pam_message **msg,
						struct pam_response **resp, void *appdata_ptr);
string findValidRandomTheme(const string& set);
void HandleSignal(int sig);

// I really didn't wanna put these globals here, but it's the only way...
Display* dpy;
int scr;
Window win;
Window root;
Cfg* cfg;
Panel* loginPanel;
string themeName = "";
string themedir = "";
const char *display = getenv("DISPLAY");

pam_handle_t *pam_handle;
struct pam_conv conv = {ConvCallback, NULL};

CARD16 dpms_standby, dpms_suspend, dpms_off, dpms_level;
BOOL dpms_state, using_dpms;
int term;

static void
die(const char *errstr, ...) {
	va_list ap;

	va_start(ap, errstr);
	vfprintf(stderr, errstr, ap);
	va_end(ap);
	exit(EXIT_FAILURE);
}

int main(int argc, char **argv) {
        int error;
  
	if((argc == 2) && !strcmp("-v", argv[1]))
		die(APPNAME"-"VERSION", © 2010-2012 Joel Burget\n");
	else if(argc != 1)
		die("usage: "APPNAME" [-v]\n");

	void (*prev_fn)(int);

	// restore DPMS settings should slimlock be killed in the line of duty
	prev_fn = signal(SIGTERM, HandleSignal);
	if (prev_fn == SIG_IGN) signal(SIGTERM, SIG_IGN);

	// create a lock file to solve mutliple instances problem
	// /var/lock used to be the place to put this, now it's /run/lock
	// ...i think
	struct stat statbuf;
	int lock_file;

	// try /run/lock first, since i believe it's preferred
	if (!stat("/run/lock", &statbuf))
		lock_file = open("/run/lock/"APPNAME".lock", O_CREAT | O_RDWR, 0666);
	else
		lock_file = open("/var/lock/"APPNAME".lock", O_CREAT | O_RDWR, 0666);

	int rc = flock(lock_file, LOCK_EX | LOCK_NB);

	if(rc) {
		if(EWOULDBLOCK == errno)
			die(APPNAME" already running\n");
	}

	unsigned int cfg_passwd_timeout;
	// Read user's current theme
	cfg = new Cfg;
	cfg->readConf(CFGFILE);
	cfg->readConf(SLIMLOCKCFG);
	string themebase = "";
	string themefile = "";
	themeName = "";
	themebase = string(THEMESDIR) + "/";
	themeName = cfg->getOption("current_theme");
	string::size_type pos;
	if ((pos = themeName.find(",")) != string::npos) {
		themeName = findValidRandomTheme(themeName);
	}

	bool loaded = false;
	while (!loaded) {
		themedir =  themebase + themeName;
		themefile = themedir + THEMESFILE;
		if (!cfg->readConf(themefile)) {
			if (themeName == "default") {
				cerr << APPNAME << ": Failed to open default theme file "
					 << themefile << endl;
				exit(ERR_EXIT);
			} else {
				cerr << APPNAME << ": Invalid theme in config: "
					 << themeName << endl;
				themeName = "default";
			}
		} else {
			loaded = true;
		}
	}

	display = getenv("DISPLAY");
	if (!display)
		display = DISPLAY;

  	if(!(dpy = XOpenDisplay(display)))
		die(APPNAME": cannot open display\n");

	root = RootWindow(dpy, scr);
	InitWindow();
	
	// disable tty switching
	if(cfg->getOption("tty_lock") == "1") {
		if ((term = open("/dev/console", O_RDWR)) == -1)
			perror("error opening console");

		if ((ioctl(term, VT_LOCKSWITCH)) == -1)
			perror("error locking console");
	}

	// Get password timeout
	cfg_passwd_timeout = Cfg::string2int(cfg->getOption("wrong_passwd_timeout").c_str());
	// Let's just make sure it has a sane value
	cfg_passwd_timeout = cfg_passwd_timeout > 60 ? 60 : cfg_passwd_timeout;

	// Main loop
	while (true)
	{
	        if (loginPanel->MonitorsChanged()) {
   		        loginPanel->ClosePanel();
			delete loginPanel;
			CloseWindow();
			InitWindow();
		}

		loginPanel->ResetPasswd();

		// AuthenticateUser returns true if authenticated
		if (AuthenticateUser())
			break;

		if (!loginPanel->MonitorsChanged())
		        loginPanel->WrongPassword(cfg_passwd_timeout);
	}

	loginPanel->ClosePanel();
	delete loginPanel;

	
	CloseWindow();
	// Get DPMS stuff back to normal
	if (using_dpms) {
	  DPMSSetTimeouts(dpy, dpms_standby, dpms_suspend, dpms_off);
	  // turn off DPMS if it was off when we entered
		if (!dpms_state)
		  DPMSDisable(dpy);
	}

	XUnmapWindow(dpy, root);
	XDestroyWindow(dpy, root);
	XCloseDisplay(dpy);
	close(lock_file);

	if(cfg->getOption("tty_lock") == "1") {
		if ((ioctl(term, VT_UNLOCKSWITCH)) == -1) {
			perror("error unlocking console");
		}
	}
	close(term);

	delete cfg;
	return 0;
}

void InitWindow()
{
  	scr = DefaultScreen(dpy);

	XSetWindowAttributes wa;
	wa.override_redirect = 1;
	wa.background_pixel = BlackPixel(dpy, scr);

	XFlush(dpy);
	// Create a full screen window
	win = XCreateWindow(dpy,
	  root,
	  0,
	  0,
	  DisplayWidth(dpy, scr),
	  DisplayHeight(dpy, scr),
	  0,
	  DefaultDepth(dpy, scr),
	  CopyFromParent,
	  DefaultVisual(dpy, scr),
	  CWOverrideRedirect | CWBackPixel,
	  &wa);
	XMapWindow(dpy, win);

	for (int len = 1000; len; len--) {
		if(XGrabKeyboard(dpy, root, True, GrabModeAsync, GrabModeAsync, CurrentTime)
			== GrabSuccess)
			break;
		usleep(1000);
	}
	XSelectInput(dpy, win, ExposureMask | KeyPressMask);

	// This hides the cursor if the user has that option enabled in their
	// configuration
	HideCursor();

	loginPanel = new Panel(dpy, scr, win, cfg, themedir, Panel::Mode_Lock);
	int ret = pam_start(APPNAME, loginPanel->GetName().c_str(), &conv, &pam_handle);
	// If we can't start PAM, just exit because slimlock won't work right
	if (ret != PAM_SUCCESS)
		die("PAM: %s\n", pam_strerror(pam_handle, ret));

	// Set up DPMS
	unsigned int cfg_dpms_standby, cfg_dpms_off;
	cfg_dpms_standby = Cfg::string2int(cfg->getOption("dpms_standby_timeout").c_str());
	cfg_dpms_off = Cfg::string2int(cfg->getOption("dpms_off_timeout").c_str());
	using_dpms = DPMSCapable(dpy) && (cfg_dpms_standby > 0);
	if (using_dpms) {
		DPMSGetTimeouts(dpy, &dpms_standby, &dpms_suspend, &dpms_off);

		DPMSSetTimeouts(dpy, cfg_dpms_standby,
						cfg_dpms_standby, cfg_dpms_off);

		DPMSInfo(dpy, &dpms_level, &dpms_state);
		if (!dpms_state)
			DPMSEnable(dpy);
	}

	return;
}

void CloseWindow()
{        
        pam_end(pam_handle, 0);
	XFlush(dpy);	
}

void HideCursor()
{
	if (cfg->getOption("hidecursor") == "true") {
		XColor black;
		char cursordata[1];
		Pixmap cursorpixmap;
		Cursor cursor;
		cursordata[0] = 0;
		cursorpixmap = XCreateBitmapFromData(dpy, win, cursordata, 1, 1);
		black.red = 0;
		black.green = 0;
		black.blue = 0;
		cursor = XCreatePixmapCursor(dpy, cursorpixmap, cursorpixmap,
									 &black, &black, 0, 0);
		XFreePixmap(dpy, cursorpixmap);
		XDefineCursor(dpy, win, cursor);
	}
}

static int ConvCallback(int num_msgs, const struct pam_message **msg,
						struct pam_response **resp, void *appdata_ptr)
{
	loginPanel->EventHandler(Panel::Get_Passwd);
	if (loginPanel->MonitorsChanged()) {
	        return PAM_ABORT;
	}

	// PAM expects an array of responses, one for each message
	if (num_msgs == 0 ||
		(*resp = (pam_response*) calloc(num_msgs, sizeof(struct pam_message))) == NULL)
		return PAM_BUF_ERR;

	for (int i = 0; i < num_msgs; i++) {
		if (msg[i]->msg_style != PAM_PROMPT_ECHO_OFF &&
			msg[i]->msg_style != PAM_PROMPT_ECHO_ON)
			continue;

		// return code is currently not used but should be set to zero
		resp[i]->resp_retcode = 0;
		if ((resp[i]->resp = strdup(loginPanel->GetPasswd().c_str())) == NULL) {
			free(*resp);
			return PAM_BUF_ERR;
		}
	}

	return PAM_SUCCESS;
}

bool AuthenticateUser()
{
	return(pam_authenticate(pam_handle, 0) == PAM_SUCCESS);
}

string findValidRandomTheme(const string& set)
{
	// extract random theme from theme set; return empty string on error
	string name = set;
	struct stat buf;

	if (name[name.length() - 1] == ',') {
		name.erase(name.length() - 1);
	}

	Util::srandom(Util::makeseed());

	vector<string> themes;
	string themefile;
	Cfg::split(themes, name, ',');
	do {
		int sel = Util::random() % themes.size();

		name = Cfg::Trim(themes[sel]);
		themefile = string(THEMESDIR) +"/" + name + THEMESFILE;
		if (stat(themefile.c_str(), &buf) != 0) {
			themes.erase(find(themes.begin(), themes.end(), name));
			cerr << APPNAME << ": Invalid theme in config: "
				 << name << endl;
			name = "";
		}
	} while (name == "" && themes.size());
	return name;
}

void HandleSignal(int sig)
{
	// Get DPMS stuff back to normal
	if (using_dpms) {
		DPMSSetTimeouts(dpy, dpms_standby, dpms_suspend, dpms_off);
		// turn off DPMS if it was off when we entered
		if (!dpms_state)
			DPMSDisable(dpy);
	}

	if ((ioctl(term, VT_UNLOCKSWITCH)) == -1) {
		perror("error unlocking console");
	}
	close(term);

	loginPanel->ClosePanel();
	delete loginPanel;

	die(APPNAME": Caught signal; dying\n");
}


cfg.cpp
/* SLiM - Simple Login Manager
   Copyright (C) 2004-06 Simone Rota <sip@varlock.com>
   Copyright (C) 2004-06 Johannes Winkelmann <jw@tks6.net>
   Copyright (C) 2012-13 Nobuhiro Iwamatsu <iwamatsu@nigauri.org>

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2 of the License, or
   (at your option) any later version.
*/

#include <fstream>
#include <string>
#include <iostream>
#include <unistd.h>
#include <stdlib.h>

#include <sys/types.h>
#include <sys/stat.h>
#include <dirent.h>

#include "cfg.h"

using namespace std;

typedef pair<string,string> option;

Cfg::Cfg()
	: currentSession(-1)
{
	/* Configuration options */
	options.insert(option("default_path","/bin:/usr/bin:/usr/local/bin"));
	options.insert(option("default_xserver","/usr/bin/X"));
	options.insert(option("xserver_arguments",""));
	options.insert(option("numlock",""));
	options.insert(option("daemon",""));
	options.insert(option("xauth_path","/usr/bin/xauth"));
	options.insert(option("login_cmd","exec /bin/bash -login ~/.xinitrc %session"));
	options.insert(option("halt_cmd","/sbin/shutdown -h now"));
	options.insert(option("reboot_cmd","/sbin/shutdown -r now"));
	options.insert(option("suspend_cmd",""));
	options.insert(option("sessionstart_cmd",""));
	options.insert(option("sessionstop_cmd",""));
	options.insert(option("console_cmd","/usr/bin/xterm -C -fg white -bg black +sb -g %dx%d+%d+%d -fn %dx%d -T ""Console login"" -e /bin/sh -c ""/bin/cat /etc/issue; exec /bin/login"""));
	options.insert(option("screenshot_cmd","import -window root /slim.png"));
	options.insert(option("welcome_msg","Welcome to %host"));
	options.insert(option("session_msg","Session:"));
	options.insert(option("default_user",""));
	options.insert(option("focus_password","no"));
	options.insert(option("auto_login","no"));
	options.insert(option("current_theme","default"));
	options.insert(option("lockfile","/var/run/slim.lock"));
	options.insert(option("logfile","/var/log/slim.log"));
	options.insert(option("authfile","/var/run/slim.auth"));
	options.insert(option("shutdown_msg","The system is halting..."));
	options.insert(option("reboot_msg","The system is rebooting..."));
	options.insert(option("sessiondir",""));
	options.insert(option("hidecursor","false"));

	/* Theme stuff */
	options.insert(option("input_panel_x","50%"));
	options.insert(option("input_panel_y","40%"));
	options.insert(option("input_name_x","200"));
	options.insert(option("input_name_y","154"));
	options.insert(option("input_pass_x","-1")); /* default is single inputbox */
	options.insert(option("input_pass_y","-1"));
	options.insert(option("input_font","Verdana:size=11"));
	options.insert(option("input_color", "#000000"));
	options.insert(option("input_cursor_height","20"));
	options.insert(option("input_maxlength_name","20"));
	options.insert(option("input_maxlength_passwd","20"));
	options.insert(option("input_shadow_xoffset", "0"));
	options.insert(option("input_shadow_yoffset", "0"));
	options.insert(option("input_shadow_color","#FFFFFF"));

	options.insert(option("welcome_font","Verdana:size=14"));
	options.insert(option("welcome_color","#FFFFFF"));
	options.insert(option("welcome_x","-1"));
	options.insert(option("welcome_y","-1"));
	options.insert(option("welcome_shadow_xoffset", "0"));
	options.insert(option("welcome_shadow_yoffset", "0"));
	options.insert(option("welcome_shadow_color","#FFFFFF"));

	options.insert(option("intro_msg",""));
	options.insert(option("intro_font","Verdana:size=14"));
	options.insert(option("intro_color","#FFFFFF"));
	options.insert(option("intro_x","-1"));
	options.insert(option("intro_y","-1"));

	options.insert(option("background_style","stretch"));
	options.insert(option("background_color","#CCCCCC"));

	options.insert(option("username_font","Verdana:size=12"));
	options.insert(option("username_color","#FFFFFF"));
	options.insert(option("username_x","-1"));
	options.insert(option("username_y","-1"));
	options.insert(option("username_msg","Please enter your username"));
	options.insert(option("username_shadow_xoffset", "0"));
	options.insert(option("username_shadow_yoffset", "0"));
	options.insert(option("username_shadow_color","#FFFFFF"));

	options.insert(option("password_x","-1"));
	options.insert(option("password_y","-1"));
	options.insert(option("password_msg","Please enter your password"));

	options.insert(option("msg_color","#FFFFFF"));
	options.insert(option("msg_font","Verdana:size=16:bold"));
	options.insert(option("msg_x","40"));
	options.insert(option("msg_y","40"));
	options.insert(option("msg_shadow_xoffset", "0"));
	options.insert(option("msg_shadow_yoffset", "0"));
	options.insert(option("msg_shadow_color","#FFFFFF"));

	options.insert(option("session_color","#FFFFFF"));
	options.insert(option("session_font","Verdana:size=16:bold"));
	options.insert(option("session_x","50%"));
	options.insert(option("session_y","90%"));
	options.insert(option("session_shadow_xoffset", "0"));
	options.insert(option("session_shadow_yoffset", "0"));
	options.insert(option("session_shadow_color","#FFFFFF"));

	// slimlock-specific options
	options.insert(option("dpms_standby_timeout", "60"));
	options.insert(option("dpms_off_timeout", "600"));
	options.insert(option("wrong_passwd_timeout", "2"));
	options.insert(option("passwd_feedback_x", "50%"));
	options.insert(option("passwd_feedback_y", "10%"));
	options.insert(option("passwd_feedback_msg", "Authentication failed"));
	options.insert(option("passwd_feedback_capslock", "Authentication failed (CapsLock is on)"));
	options.insert(option("show_username", "1"));
	options.insert(option("show_welcome_msg", "0"));
	options.insert(option("tty_lock", "0"));
	options.insert(option("bell", "1"));

	error = "";
}

Cfg::~Cfg() {
	options.clear();
}
/*
 * Creates the Cfg object and parses
 * known options from the given configfile / themefile
 */
bool Cfg::readConf(string configfile) {
	int n = -1;
	size_t pos = 0;
	string line, next, op, fn(configfile);
	map<string,string>::iterator it;
	ifstream cfgfile(fn.c_str());

	if (!cfgfile) {
		error = "Cannot read configuration file: " + configfile;
		return false;
	}
	while (getline(cfgfile, line)) {
		if ((pos = line.find('\\')) != string::npos) {
			if (line.length() == pos + 1) {
				line.replace(pos, 1, " ");
				next = next + line;
				continue;
			} else
				line.replace(pos, line.length() - pos, " ");
		}

		if (!next.empty()) {
			line = next + line;
			next = "";
		}
		it = options.begin();
		while (it != options.end()) {
			op = it->first;
			n = line.find(op);
			if (n == 0)
				options[op] = parseOption(line, op);
			++it;
		}
	}
	cfgfile.close();

	fillSessionList();

	return true;
}

/* Returns the option value, trimmed */
string Cfg::parseOption(string line, string option ) {
	return Trim( line.substr(option.size(), line.size() - option.size()));
}

const string& Cfg::getError() const {
	return error;
}

string& Cfg::getOption(string option) {
	return options[option];
}

/* return a trimmed string */
string Cfg::Trim( const string& s ) {
	if ( s.empty() ) {
		return s;
	}
	int pos = 0;
	string line = s;
	int len = line.length();
	while ( pos < len && isspace( line[pos] ) ) {
		++pos;
	}
	line.erase( 0, pos );
	pos = line.length()-1;
	while ( pos > -1 && isspace( line[pos] ) ) {
		--pos;
	}
	if ( pos != -1 ) {
		line.erase( pos+1 );
	}
	return line;
}

/* Return the welcome message with replaced vars */
string Cfg::getWelcomeMessage(){
	string s = getOption("welcome_msg");
	int n = s.find("%host");
	if (n >= 0) {
		string tmp = s.substr(0, n);
		char host[40];
		gethostname(host,40);
		tmp = tmp + host;
		tmp = tmp + s.substr(n+5, s.size() - n);
		s = tmp;
	}
	n = s.find("%domain");
	if (n >= 0) {
		string tmp = s.substr(0, n);;
		char domain[40];
		getdomainname(domain,40);
		tmp = tmp + domain;
		tmp = tmp + s.substr(n+7, s.size() - n);
		s = tmp;
	}
	return s;
}

int Cfg::string2int(const char* string, bool* ok) {
	char* err = 0;
	int l = (int)strtol(string, &err, 10);
	if (ok) {
		*ok = (*err == 0);
	}
	return (*err == 0) ? l : 0;
}

int Cfg::getIntOption(std::string option) {
	return string2int(options[option].c_str());
}

/* Get absolute position */
int Cfg::absolutepos(const string& position, int max, int width) {
	int n = position.find("%");
	if (n>0) { /* X Position expressed in percentage */
		int result = (max*string2int(position.substr(0, n).c_str())/100) - (width / 2);
		return result < 0 ? 0 : result ;
	} else { /* Absolute X position */
		return string2int(position.c_str());
	}
}

/* split a comma separated string into a vector of strings */
void Cfg::split(vector<string>& v, const string& str, char c, bool useEmpty) {
	v.clear();
	string::const_iterator s = str.begin();
	string tmp;
	while (true) {
		string::const_iterator begin = s;
		while (*s != c && s != str.end()) { ++s; }
		tmp = string(begin, s);
		if (useEmpty || tmp.size() > 0)
			v.push_back(tmp);
		if (s == str.end()) {
			break;
		}
		if (++s == str.end()) {
			if (useEmpty)
				v.push_back("");
			break;
		}
	}
}

void Cfg::fillSessionList(){
	string strSessionList = getOption("sessions");
	string strSessionDir  = getOption("sessiondir");

	sessions.clear();

	if( !strSessionDir.empty() ) {
		DIR *pDir = opendir(strSessionDir.c_str());

		if (pDir != NULL) {
			struct dirent *pDirent = NULL;

			while ((pDirent = readdir(pDir)) != NULL) {
				string strFile(strSessionDir);
				strFile += "/";
				strFile += pDirent->d_name;

				struct stat oFileStat;

				if (stat(strFile.c_str(), &oFileStat) == 0) {
					if (S_ISREG(oFileStat.st_mode) &&
						access(strFile.c_str(), R_OK) == 0){
							ifstream desktop_file( strFile.c_str() );
							if (desktop_file){
								string line, session_name = "", session_exec = "";
								while (getline( desktop_file, line )) {
								if (line.substr(0, 5) == "Name=") {
									session_name = line.substr(5);
									if (!session_exec.empty()) break;
								} else if (line.substr(0, 5) == "Exec=") {
									session_exec = line.substr(5);
									if (!session_name.empty()) break;
								}
							}
							desktop_file.close();
							if (!session_name.empty() && !session_exec.empty()) {
								pair<string,string> session(session_name,session_exec);
								sessions.push_back(session);
							} else if (access(strFile.c_str(), X_OK) == 0) {
								pair<string,string> session(string(pDirent->d_name),strFile);
								sessions.push_back(session);
							}
						}
					}
				}
			}
			closedir(pDir);
		}
	}

	if (sessions.empty()){
		if (strSessionList.empty()) {
			pair<string,string> session("","");
			sessions.push_back(session);
		} else {
			// iterate through the split of the session list
			vector<string> sessit;
			split(sessit,strSessionList,',',false);
			for (vector<string>::iterator it = sessit.begin(); it != sessit.end(); ++it) {
				pair<string,string> session(*it,*it);
				sessions.push_back(session);
			}
		}
	}
}

pair<string,string> Cfg::nextSession() {
	currentSession = (currentSession + 1) % sessions.size();
	return sessions[currentSession];
}
